### ISSUE TITLE
Refactor write_iteration* functions

### ISSUE THREAD

**Issue by dustinvtran** [stan-dev/stan#1538]
[`write_iteration_csv(...)`](https://github.com/stan-dev/stan/blob/develop/src/stan/services/io/write_iteration_csv.hpp) currently requires specifying the log probability `lp`. This is unneeded for variational inference, and as a hack currently writes 0.0 there: https://github.com/stan-dev/stan/blob/develop/src/stan/variational/advi.hpp#L315-L317

From [stan-users mailing list](https://groups.google.com/forum/#!topic/stan-dev/Bo5fWHI12iA):
```
Let's look into refactoring that into less of a hack.  We shouldn't be requiring
lp__ in anything downstream.   Feel free to bring these things up during
design, and we can look into redoing them as we go so we don't wind up with
a hack release that we want to fix later by getting rid of lp__.

- Bob
```

**Comment by syclik** [comment]
Thanks for hunting this down. We should definitely generalize our
functions.

**Comment by betanalpha** [comment]
Cough, cough — already done in the writer refactor — cough, cough

**Comment by syclik** [comment]
Are you sure we can't break that refactoring into a lot of smaller pull
requests?
Cough, cough — already done in the writer refactor — cough, cough

On Jul 10, 2015, at 10:56 PM, Daniel Lee <notifications@github.com> wrote:

> Thanks for hunting this down. We should definitely generalize our
> functions.
> On Jul 10, 2015 4:43 PM, "Dustin Tran" <notifications@github.com> wrote:
>
> > write_iteration_csv(...)
> > <
https://github.com/stan-dev/stan/blob/develop/src/stan/services/io/write_iteration_csv.hpp
>
> > currently requires specifying the log probability lp. This is unneeded
> > for variational inference, and as a hack currently writes 0.0 there:
> >
https://github.com/stan-dev/stan/blob/develop/src/stan/variational/advi.hpp#L315-L317
> >
> > From stan-users mailing list
> > <https://groups.google.com/forum/#!topic/stan-dev/Bo5fWHI12iA>:
> >
> > Let's look into refactoring that into less of a hack. We shouldn't be
requiring
> > lp__ in anything downstream. Feel free to bring these things up during
> > design, and we can look into redoing them as we go so we don't wind up
with
> > a hack release that we want to fix later by getting rid of lp__.
> >
> > - Bob
> >
> > —
> > Reply to this email directly or view it on GitHub
> > <https://github.com/stan-dev/stan/issues/1538>.
> >
> —
> Reply to this email directly or view it on GitHub.
>

—
Reply to this email directly or view it on GitHub
<https://github.com/stan-dev/stan/issues/1538#issuecomment-120524415>.

**Comment by dustinvtran** [comment]
Is this issue solved yet? I think so but I don't want to close it without confirmation.

**Comment by syclik** [comment]
I don't think so... it's not updated on develop, right? (Normally, I'd check first, but hopefully it's easy for you to confirm.)

**Comment by dustinvtran** [comment]
Oh, you're right. I mistook my above issue with something else about streams. Yes we still have to do a hack by inputting 0 for the log prob's.

**Comment by akucukelbir** [comment]
is michael's refactor comment (betanalpha commented on Jul 10, 2015) relevant?

or do we need to make some further change to `write_iteration`?

**Comment by martinmodrak** [comment]
This seems to be already done - lp is not written by ADVI (by my reading of the codebase) could @betanalpha  verify and close?

**Comment by bob-carpenter** [comment]
Thanks, @martinmodrak---this kind of purposeful auditing of issues is really helpful.

**Comment by mcol** [comment]
I think it's not done, as `src/stan/variational/advi.hpp` still contains these lines:
```
// Write lp__, log_p, and log_g.
values.insert(values.begin(), {0, log_p, log_g});
```
